Guidance hint solution that resolves #2007

jd-alexander opened the pull request:
Closes #2007 

#### What has been done to verify that this works as intended?

To verify that this functionality works I manually tested it by creating a form that has guidance attached to the hint of its question. Once that was done, I loaded the form into the Collect app, enabled guidance hints to be shown on questions within the Form Management Settings and opened the form's question which displayed the hint. 

To further verify this functionality I wrote an Espresso test that enabled the guidance setting and I created a test instance of `FormEntryActivity` with the guidance hint form. I then fetched the guidance hint from the respective `FormEntryPrompt` and compared it with the text of the TextView on the screen that's supposed to perform the rendering. 

#### Why is this the best possible solution? Were any other approaches considered?
This is the best possible solution due to the nature of how the JavaRosa SDK allows text strings with form attributes to be fetched. 

#### Are there any risks to merging this code? If so, what are they?
The only risks associated with merging this code would be the changes that were made to the `HelpTextView`'s layout constraints that were associated with the rendering of the `AnswerView.` This was being done entirely in code so a xml layout file was used to encapsulate the additional layout that was necessary for the guidance text view. Nonetheless, I made several checks to ensure that these changes didn't break the functionality of the questions and the preexisting UI tests along with the UI test that was written for this feature verifies that this won't cause an impact.

#### Do we need any specific form for testing your changes? If so, please attach one.
To test this functionality you need a form that has a [guidance hint](https://opendatakit.github.io/xforms-spec/#languages) eg. 
[guidance_hint_form.xml.txt](https://github.com/opendatakit/collect/files/1888351/guidance_hint_form.xml.txt)



#### Does this change require updates to documentation? If so, please file an issue at https://github.com/opendatakit/docs/issues/new and include the link below.
https://github.com/opendatakit/docs/issues/635

jd-alexander committed:
created a layout and view binding mechanisms for all help/hint related constructs. This will allow for the hint and the hint's guidance to live within the same layout. Also added a key to preferences that will be used to store the guidance setting.

jd-alexander committed:
added list preference for guidance setting

jd-alexander committed:
fetched guidance hint from the associated FormEntryCaption based on the hint's text id and form type of guidance.

jd-alexander committed:
added animations to collapse guidance setup.

jd-alexander committed:
wrote integration tests to verify that guidance hint works.

jd-alexander committed:
performed cleanup of test.

jd-alexander committed:
added styling and configurations for rendering to guidance text.

jd-alexander committed:
re-enabled D8.

jd-alexander committed:
removed this visibility check on help related layouts because if the help text is already gone then it's parent layout won't have a height due to it having the WRAP_CONTENT property so the AnswerView will automatically be positioned beneath the previous layout.

jd-alexander committed:
resolved pmd issues.

jd-alexander committed:
fixed checkstyle issues

jd-alexander committed:
performed several fixes to adhere to codacy styles

jd-alexander committed:
changed assertTrue to assertFalse for string null check.

jd-alexander committed:
added view id matcher static import.

lognaturel commented on collect_app/src/main/res/layout/help_text_layout.xml:
```diff
@@ -0,0 +1,24 @@
+    <TextView
+        android:id="@+id/help_text_view"
+        tools:text="Testing this out."
+        android:gravity="center"
```
This makes all hint text that wraps centered which doesn't look right. See All Widgets form to see what I mean!

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/widgets/QuestionWidget.java:
```diff
@@ -592,4 +677,6 @@ public MediaPlayer getPlayer() {
     public int getPlayColor() {
         return playColor;
     }
+
```
There shouldn't be any line breaks here.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/widgets/QuestionWidget.java:
```diff
@@ -309,12 +383,18 @@ protected void addHelpTextView(View v) {
-        TextView helpText = new TextView(getContext());
+    private View createHelpTextLayout() {
+        return LayoutInflater.from(getContext()).inflate(R.layout.help_text_layout, null);
+
```
Please remove extra line breaks! I saw a few other places where there were extra ones so would be good to double check the diff.

jd-alexander replied:
No prob :) will definitely keep an eye out for them!

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/widgets/QuestionWidget.java:
```diff
@@ -115,13 +127,73 @@ public boolean onError(MediaPlayer mp, int what, int extra) {
+    }
+
+    private TextView setupGuidanceTextAndLayout(TextView guidanceTextView, FormEntryPrompt prompt) {
+
```
Our style doesn't call for a line break here.

lognaturel commented on collect_app/src/main/res/values/strings.xml:
```diff
@@ -584,4 +587,5 @@
     <string name="start_video_capture_instruction">Tap the screen to start recording</string>
     <string name="stop_video_capture_instruction">Recording started. Tap again to stop</string>
     <string name="no_columns_to_upload">No columns found in the form to upload.</string>
+    <string name="guidance_hint_title">Show Guidance For Questions</string>
```
Should be in sentence case ("Show guidance for questions").

lognaturel commented on collect_app/src/main/res/values/strings.xml:
```diff
@@ -495,6 +495,9 @@
     <string name="autosend_selector_title">Auto send</string>
     <string name="off">Off</string>
+    <string name="guidance_no">No</string>
+    <string name="guidance_yes">Yes - Always Shown</string>
```
always shown (all lowercase)

lognaturel commented on collect_app/src/main/res/values/strings.xml:
```diff
@@ -495,6 +495,9 @@
     <string name="off">Off</string>
+    <string name="guidance_no">No</string>
+    <string name="guidance_yes">Yes - Always Shown</string>
+    <string name="guidance_yes_collapsed">Yes - Collapsed</string>
```
collapsed (all lowercase)

lognaturel reviewed (commented):
Looking great to me! I've marked a few quick things inline as I saw them but haven't done a full review yet. 

Let's wait a day or two for feedback about the UI: https://forum.opendatakit.org/t/collect-collapsible-training-hints-guidance/12247/3

I find the (i) icon very difficult to hit even on a Galaxy Tab A tablet. Is it possible only the actual path is tappable and not the transparent parts? Could we perhaps expand the tappable area?

I wasn't sure initially that I liked the icon but I haven't come up with anything better so it has grown on me.

jd-alexander commented:
Thanks for the review! 
Icon/UI - Yes I am not too sure about it either so I just put it in place until we could all decide on something together. I have an idea to spruce up the UI a bit so I am gonna make those changes and share them here. 

Tappable area 
- > I find the (i) icon very difficult to hit even on a Galaxy Tab A tablet. Is it possible only the actual path is tappable and not the transparent parts? Could we perhaps expand the tappable area?

Yes, we could make clicks only take place on the actual path. Currently I just have an `OnClickListener` on the `TextView` because I am using it's `CompoundDrawable` mechanism  so no matter the length of the hint the icon will always be properly aligned at the end of the text. Started working on a layout with a `TextView` and `ImageView` but I just thought this approach would yield the same results.

lognaturel commented:
> Currently I just have an OnClickListener on the TextView

Of course, I hadn't noticed that! That explains why I was having a tough time when I tried to tap the (i). Eventually it would be good to have both text and icon be tappable but let's wait for some preliminary UI feedback. 😊

jd-alexander commented:
Okay :) I made just the drawable clickable for now. Gonna do some thorough testing of it to see how it behaves.

jd-alexander committed:
performed style and drawable click fixes

yanokwa commented on gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -1,6 +1,7 @@
-distributionUrl=https\://services.gradle.org/distributions/gradle-4.5.1-bin.zip
-android.enableD8.desugaring=true
\ No newline at end of file
+distributionUrl=https\://services.gradle.org/distributions/gradle-4.5.1-all.zip
```
I think this change should be reverted because we don't need the sources and documentation that `all` comes with. Why download things we don't need 😂

jd-alexander replied:
haha no prob. Never knew that's what the `all` meant. Thanks for pointing that out.

yanokwa commented:
I think this works really nicely, and we have tests! Hard to do complain about that...

Some changes I'd like to see before this goes to testing.
* The `i` icon should probably be black. Gray makes it look disabled.
* The icon isn't a huge touch surface, so maybe we can make the whole hint text clickable, to make life easier for users? This is similar to how you can select the entire line for a radio button in selects ones.

I'm not a huge fan of the `i` after the text (although @lognaturel prefers it!). I had similar pushback when we added the red star for required questions (https://github.com/opendatakit/collect/pull/1828#pullrequestreview-94182756) and I was able to convince folks to move the red star to the beginning. I'm not necessarily arguing for a change, but I'm curious how you decided to put it at the end, @jd-alexander.

jd-alexander commented:
Good looking out @yanokwa . Honestly, I was used to putting asterisks/icons about hints at the end of the text from my old job so that's the reason I placed it there. I did some UX research on this topic and I found some [detailed explanations from the folks over at UX Stack Exchange](https://ux.stackexchange.com/questions/56023/when-a-button-contains-text-and-an-icon-which-should-come-first) that justifies it being placed before the text so I am gonna opt for that approach because it makes more sense and it actually lessens most of the issues I had to deal with such as the icon trailing the text and causing inconsistencies with other hint guidance's that might be shorter. I am gonna work on this change and a subtle effect to make this more appealing and share it here :smile:

yanokwa commented:
@jd-alexander If you can get those changes sent in soon and fix the conflicts, I think we'd be ready to test and merge!

jd-alexander commented:
@yanokwa Yes I am gonna finish this up! Was just a bit pre occupied with the SMS stuff but yes I want to finalize this so I will spend some time doing this tomorrow!

jd-alexander commented:
@yanokwa I was thinking of adding a background to the guidance hint when it's in collapse mode so when it gets expanded it has a bit more body and outline. I have this working example in my repo of how it would look when it's open. What do you think?
I included this in the pushes I sent so you can take a look if necessary. 

![Collapsed Hint Guidance](https://user-images.githubusercontent.com/1509205/39269038-1217ee76-48c1-11e8-840a-cadf04aa1240.png)

jd-alexander committed:
added a background to guidance hint's layout so that a noticeable outline is displayed when it's expanded in collapsed mode.

jd-alexander committed:
Merge branch 'master' into guidance_hint

jd-alexander committed:
several minor style and project fixes

jd-alexander committed:
Update colors.xml

jd-alexander committed:
Merge branch 'guidance_hint' of https://github.com/jd-alexander/collect into guidance_hint

# Conflicts:
#	collect_app/src/main/res/values/colors.xml

jd-alexander committed:
removed unused resources

jd-alexander commented:
Hey @yanokwa I submitted the changes. Not sure what's causing this build to fail though. Probably triggering a rebuild might cause it to pass because there are no style errors.

jd-alexander committed:
minor fix to trigger rebuild.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/preferences/PreferenceKeys.java:
```diff
@@ -21,6 +21,7 @@
     public static final String KEY_CONSTRAINT_BEHAVIOR      = "constraint_behavior";
     public static final String KEY_HIGH_RESOLUTION          = "high_resolution";
     public static final String KEY_IMAGE_SIZE               = "image_size";
+    public static final String KEY_GUIDANCE_HINT               = "guidance_hint";
```
Indentation here should match that of other values.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/preferences/GuidanceHint.java:
```diff
@@ -0,0 +1,28 @@
+    public static GuidanceHint get(String name) {
+        for (GuidanceHint hint : GuidanceHint.values()) {
+            if (hint.name.equals(name)) {
+
```
Remove extra blank line.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/utilities/AnimateUtils.java:
```diff
@@ -1,15 +1,23 @@
  */
 public class AnimateUtils {
 
+
```
Remove extra blank line

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/utilities/AnimateUtils.java:
```diff
@@ -39,4 +47,114 @@ public void onAnimationRepeat(Animation animation) {
         view.startAnimation(scaleInAnimation);
     }
+
+
```
Remove extra blank line

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/utilities/AnimateUtils.java:
Remove extra blank line

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/utilities/AnimateUtils.java:
Remove extra blank line

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/utilities/AnimateUtils.java:
Remove extra blank line

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/utilities/UiUtils.java:
```diff
@@ -34,4 +37,33 @@ private UiUtils() {
+
+
+    /**
+     * This method converts dp unit to equivalent pixels, depending on device density.
```
Remove "This method"

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/utilities/UiUtils.java:
```diff
@@ -34,4 +37,33 @@ private UiUtils() {
+    }
+
+    /**
+     * This method converts device specific pixels to density independent pixels.
```
Remove "This method"

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/widgets/QuestionWidget.java:
```diff
@@ -115,13 +129,76 @@ public boolean onError(MediaPlayer mp, int what, int extra) {
+
+        if (android.text.TextUtils.isEmpty(guidanceHint)) {
+            return null;
+
```
Remove extra blank line

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/widgets/QuestionWidget.java:
```diff
@@ -115,13 +129,76 @@ public boolean onError(MediaPlayer mp, int what, int extra) {
-    /** Releases resources held by this widget */
+
+    private TextView configureGuidanceTextView(TextView guidanceTextView, String guidance) {
+
```
Remove extra blank line

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/widgets/QuestionWidget.java:
```diff
@@ -115,13 +129,76 @@ public boolean onError(MediaPlayer mp, int what, int extra) {
+        guidanceTextView.setTextColor(ContextCompat.getColor(getContext(), R.color.primaryTextColor));
+        guidanceTextView.setMovementMethod(LinkMovementMethod.getInstance());
+        return guidanceTextView;
+
```
Remove extra blank line

lognaturel reviewed (commented):
Few small things and this will be ready for QA! 

- The guidance hint shouldn't disappear on rotate
- The hint font size is already small so I would like guidance hints to have the same font size

I don't find the box around the guidance hint necessary but I don't feel very strongly about it. @yanokwa?

yanokwa commented:
I don't love the box.

jd-alexander commented:
Okay no prob! Will remove the box!

lognaturel commented:
Awesome! Once you've done the changes and the merge conflicts are resolved (sorry!), you can add the `needs-testing` tag with `@opendatakit-bot label needs-testing` and it will go through the QA process.

jd-alexander committed:
added the ability to store expanded/collapsed state of guidance hint. Performed several style fixes.

jd-alexander committed:
renamed collapsed key to expanded to match property it's refernencing.Other minor fixes

jd-alexander committed:
Merge branch 'master' into guidance_hint

jd-alexander committed:
integrated guidance hint with dark/light theme functionality

jd-alexander commented:
@opendatakit-bot label needs-testing

jd-alexander commented:
- Removed box outline
- Added collapsed state to saved state
- Integrated control with dark theme
- Corrected style errors

Ready for testing :fireworks:

mmarciniak90 commented:
@jd-alexander I noticed a problem in case:
1. User installs the application (`Show guidance for questions` is set as a `No` by default)
2. User starts `Guidance Form Sample` form
3. User goes to `General Settings`
4. User changes `Show guidance for questions` to `Yes - collapsed`
5. User backs to form
6. User sees error
![2105](https://user-images.githubusercontent.com/3907554/39761828-a0708cf8-52c8-11e8-944f-cead2c90564a.gif)

Additionally, `i` icon is cut off on devices with >5 inches screen
![screenshot_2018-05-08-16-09-47](https://user-images.githubusercontent.com/3907554/39762384-0a8d037c-52ca-11e8-8400-c50e983a12de.png)

lognaturel commented:
Good catch, @mmarciniak90! @jd-alexander, hopefully these are relatively quick fixes. Would be great to get this in a beta by Thursday!

jd-alexander commented:
Thanks for catching these @mmarciniak90  @lognaturel They should be! Working on them now.

jd-alexander committed:
Refactored hint guidance layout approach

The hint guidance was being attached by using CompoundDrwables of TextView. The issue with using this was that there wasn't alot of flexibility with the layout attributes that could be applied to the drawable eg. padding can be applied (which was causing the bug) but not margins so I removed the drawable and created a layout to host the guidance textview and an an imageview for the icon

jd-alexander committed:
Merge branch 'master' into guidance_hint

jd-alexander commented:
@mmarciniak90 resolved the issues! Ready for another test to verify.

mmarciniak90 commented:
@jd-alexander mentioned problems are resolved!
I checked more question types and I think that repeating guidance for `Full date and time` question type is not expected.
![screenshot_2018-05-18-12-52-01](https://user-images.githubusercontent.com/3907554/39809223-bb662420-5370-11e8-82aa-d8a80c84da02.png)
You can use form: [hints.xml.txt](https://github.com/opendatakit/collect/files/1987313/hints.xml.txt)

jd-alexander committed:
Help layout hidden for child widgets of DateTimeWidget

The TimeWidget and DateWidget's help layout was hidden so that only the parent widget's layout is displayed. Previous only the help text was being hidden thus these changes had to take place due to the introduction of the new layout structure for hint guidance.

jd-alexander committed:
Merge remote-tracking branch 'origin/guidance_hint' into guidance_hint

jd-alexander commented:
@mmarciniak90 Resolved and ready to go!

mmarciniak90 commented:
**Tested with success!**

@jd-alexander great work! Everything looks great!

Verified on Android 4.1, 4.2, 4.4., 5.1 6.0, 7.0, 8.1

Verified cases:
* clear settings
* many types of question
* full date and time question
* many questions on one page
* dark theme

@opendatakit-bot label "behavior verified"
@opendatakit-bot unlabel "needs testing"
